src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2412,6 +2410,11 @@ Status Session::receive_download_message(const DownloadMessage& message)
     }
 
     version_type server_version = m_progress.download.server_version;
+    // If there are 2 or more changesets in this message, track whether or not the changesets have the same
+    // download_server_version to help with determining if this is a bootstrap message or not.
+    bool same_server_version =
+        message.changesets.size() > 1 && progress.download.server_version == message.changesets[0].remote_version;
+    bool after_first_changeset = false;
```

Comment:
Given the discussion on slack, I think the logic here and on line 2434 isn't quite right. What we really need to track here is whether there are any changesets with duplicate `remote_version`'s in the download message. We could do something like

```
bool has_duplicate_remote_versions = false;
std::optional<version_type> last_changeset_remote_version;
...
for (const RemoteChangeset& changeset: message.changesets) {
    if (last_changeset_remote_version == changeset.remote_version) {
        has_duplicate_remote_versions = true;
    }
    last_changeset_remote_version = changeset.remote_version;
}
```

then the check for whether this is a MoreToCome/LastInBatch/SteadyState is
```
if (batch_state == LastInBatch && !has_duplicate_remote_versions) {
   return SteadyState;
}
return batch_state;
```

Reply:
I updated some of the logic around this area and added more comments to clarify the logic around bootstrap detection.